### PR TITLE
Use connection rates satisfying the connection IPR when evaluating flow-dependent skin

### PR DESCRIPTION
**Pull request by vkip**

### CODE REVIEW

**Comment by vkip** [comment]
jenkins build this please

**Comment by vkip** [comment]
No apparent impact on the tests, improves convergence on a larger model.

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 1]
```diff
@@ -1507,13 +1507,41 @@ namespace Opm
...
+
+        const double connection_pressure = ws.perf_data.pressure[perf];
+        const double cell_pressure = getValue(intQuants.fluidState().pressure(FluidSystem::gasPhaseIdx));
+        const double drawdown = cell_pressure - connection_pressure;
+        const double invB = getValue(intQuants.fluidState().invB(FluidSystem::gasPhaseIdx));
+        const double mob_g = getValue(intQuants.mobility(FluidSystem::gasPhaseIdx)) * invB;
+        const double a = d;
+        const double b = 2*scaling/wi[gas_comp_idx];
+        const double c = -1.0*b*trans_mult*this->well_index_[perf]*mob_g*drawdown;
+
```
Comment (totto82) [inline opm/simulators/wells/WellInterface_impl.hpp:1529]: I suggest using
c = -2*scaling*mob_g*drawdown
to improve readability of the code.

**Review by totto82 — APPROVED** [review]
Thanks for the contribution. I only have minor comments.

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 2]
```diff
@@ -1507,13 +1507,41 @@ namespace Opm
...
+        const double r = std::sqrt(r2);
+        const double cQ1 = (r-b)*0.5/a;
+        const double cQ2 = -(r+b)*0.5/a;
+        const double sol1 = std::abs(a*cQ1*std::abs(cQ1) + b*cQ1 + c);
+        const double sol2 = std::abs(a*cQ2*std::abs(cQ2) + b*cQ2 + c);
+        if (std::min(sol1, sol2) > 1.0e-6) {
+            return wi;
+        }
+        const double consistent_Q = (sol1 <= sol2) ? cQ1 : cQ2;
+        wi[gas_comp_idx]  = 1.0/(1.0/(trans_mult * this->well_index_[perf]) + (std::abs(consistent_Q)/2 * d / scaling));
```
[the remark this answers is not in the document]
Comment (vkip) [reply opm/simulators/wells/WellInterface_impl.hpp:1544]: Will do, hypothesis is essentially that choosing the smallest solution here gives smallest skin, possibly enabling finding the highest-rate solution.

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 3]
```diff
@@ -1507,13 +1507,41 @@ namespace Opm
...
+        if (r2 <= 0) {
+            return wi;
+        }
+        const double r = std::sqrt(r2);
+        const double cQ1 = (r-b)*0.5/a;
+        const double cQ2 = -(r+b)*0.5/a;
+        const double sol1 = std::abs(a*cQ1*std::abs(cQ1) + b*cQ1 + c);
+        const double sol2 = std::abs(a*cQ2*std::abs(cQ2) + b*cQ2 + c);
+        if (std::min(sol1, sol2) > 1.0e-6) {
+            return wi;
```
[the remark this answers is not in the document]
Comment (vkip) [reply opm/simulators/wells/WellInterface_impl.hpp:1541]: Good question, short answer is that I was too lazy/stupid to check all solutions properly. Should of course have done that, which will ensure a solution is always found (I think). Will update.

File: opm/simulators/wells/WellInterface_impl.hpp (hunk 1 above)
Comment (vkip) [reply opm/simulators/wells/WellInterface_impl.hpp:1529]: Thanks, good point, will update.

File: opm/simulators/wells/WellInterface_impl.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (akva2) [reply opm/simulators/wells/WellInterface_impl.hpp:1541]: I might be missing but I'd say the proper lazy thing to do would be to
```c++
#include <opm/material/common/PolynomialUtils.hpp>

std::array<double,2> sol;
unsigned numsol = Opm::invertQuadraticPolynomical(sol, a, b, c);
```

[the remark this answers is not in the document]
Comment (vkip) [reply opm/simulators/wells/WellInterface_impl.hpp]: Agreed - just stupid then :-D

**Comment by vkip** [comment]
jenkins build this please

**Comment by totto82** [comment]
@vkip Do you plan to update the PR with the suggestions from @akva2 ? Or should we merge it as is?

**Comment by vkip** [comment]
> @vkip Do you plan to update the PR with the suggestions from @akva2 ? Or should we merge it as is?

Since the code is already written, tested and avoids any potential overhead I vote keep as is, but will of course update if anyone insists.

**Pull request merged**